From 45fea81826a9e4b41b073dda79537dd94bbf2e1e Mon Sep 17 00:00:00 2001
From: Ravi B <ravibabu@ti.com>
Date: Mon, 6 Dec 2010 18:31:08 +0530
Subject: [PATCH 0486/1092] musb: select fifo_mode at runtime

OMAP35x han an errata restricting active endpoints to use either
first 8K or next 8K of FIFO space while this issue has been
fixed in OMAP3630.

This issue requires the OMAP35x platform to use fifo_mode=5 but
OMAP3630 and AM35x can use fifo_mode=4 which utilises entire 16K
of FIFO.

Signed-off-by: Ajay Kumar Gupta <ajay.gupta@ti.com>
Signed-off-by: Sriramakrishnan A G <srk@ti.com>
Signed-off-by: Vaibhav Hiremath <hvaibhav@ti.com>
---
 arch/arm/mach-omap2/usb-musb.c | 9 +++++++++
 drivers/usb/musb/musb_core.c   | 4 +++-
 include/linux/usb/musb.h       | 1 +
 3 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/arch/arm/mach-omap2/usb-musb.c b/arch/arm/mach-omap2/usb-musb.c
index 8d5ed77..8387501 100644
--- a/arch/arm/mach-omap2/usb-musb.c
+++ b/arch/arm/mach-omap2/usb-musb.c
@@ -34,6 +34,7 @@
 #include "mux.h"
 
 static struct musb_hdrc_config musb_config = {
+	.fifo_mode	= 4,
 	.multipoint	= 1,
 	.dyn_fifo	= 1,
 	.num_eps	= 16,
@@ -90,6 +91,14 @@ void __init usb_musb_init(struct omap_musb_board_data *musb_board_data)
 	musb_plat.mode = board_data->mode;
 	musb_plat.extvbus = board_data->extvbus;
 
+	/*
+	 * OMAP3630/AM35x platform has MUSB RTL-1.8 which has the fix for the
+	 * issue restricting active endpoints to use first 8K of FIFO space.
+	 * This issue restricts OMAP35x platform to use fifo_mode '5'.
+	 */
+	if (cpu_is_omap3430())
+		musb_config.fifo_mode = 5;
+
 	if (cpu_is_omap3517() || cpu_is_omap3505()) {
 		oh_name = "am35x_otg_hs";
 		name = "musb-am35x";
diff --git a/drivers/usb/musb/musb_core.c b/drivers/usb/musb/musb_core.c
index 754cf11..88e661e 100644
--- a/drivers/usb/musb/musb_core.c
+++ b/drivers/usb/musb/musb_core.c
@@ -1235,7 +1235,9 @@ static int __devinit ep_config_from_table(struct musb *musb)
 	int			offset;
 	struct musb_hw_ep	*hw_ep = musb->endpoints;
 
-	if (musb->config->fifo_cfg) {
+	if (musb->config->fifo_mode) {
+		fifo_mode = musb->config->fifo_mode;
+	} else if (musb->config->fifo_cfg) {
 		cfg = musb->config->fifo_cfg;
 		n = musb->config->fifo_cfg_size;
 		goto done;
diff --git a/include/linux/usb/musb.h b/include/linux/usb/musb.h
index eb50525..2c59816 100644
--- a/include/linux/usb/musb.h
+++ b/include/linux/usb/musb.h
@@ -62,6 +62,7 @@ struct musb_hdrc_eps_bits {
 struct musb_hdrc_config {
 	struct musb_fifo_cfg	*fifo_cfg;	/* board fifo configuration */
 	unsigned		fifo_cfg_size;	/* size of the fifo configuration */
+	unsigned short		fifo_mode;	/* fifo mode to be selected */
 
 	/* MUSB configuration-specific details */
 	unsigned	multipoint:1;	/* multipoint device */
-- 
1.7.11.2

